### ISSUE TITLE
Fix false positives for punctuation rules and non-standard syntax

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#4752]
> Clearly describe the bug

There are false positives in the punctuation rules when a syntax parses a standard construct correctly, but the punctuation of the construct is non-standard.

For example, postcss-css-in-js will correctly parse the object pair `{{ color: "red" }}` as a declaration but the punctuation is non-standard and will trip up rules like `declaration-block-trailing-semicolon`.

See https://github.com/stylelint/stylelint/pull/4749.

> Which rule, if any, is the bug related to?

All punctuation rules, including those for:

- semicolon
- braces

See [this section of the standard-config docs](https://github.com/stylelint/stylelint-config-standard#using-the-config-with-sugarss-syntax) for a more comprehensive list.

> What did you expect to happen?

No false positives.

We need a set of `isStandardPunctuation*` utils. The set would work in conjunction with the preexisting  `isStandardSyntax*` utils.

The `isStandardSyntax*` utils are used when a syntax parses a non-standard construct as a standard one, e.g. postcss-scss will parse a SCSS map as a `decl` node. 

These utils have served us well. Built-in rule authors can don't have to think about non-standard syntax, and plugin authors can target non-standard syntax confident that the built-in rules won't be in conflict. For example, [dollar variables](https://github.com/kristerkari/stylelint-scss#-variable) and [map](https://github.com/kristerkari/stylelint-scss#map) rules within the [stylelint-scss plugin](https://github.com/kristerkari/stylelint-scss).

The proposed `isStandardPunctuation*` utils will bring similar benefits to built-in rules that deal with punctuation. 

We have historically said that it's the responsibility of the configuration author to turn off rules that are incompatible with their chosen syntax. However, I now think that this is:

- at odds with our `isStandardSyntax*` utils approach
- no longer viable as source files can contain more than one syntax

For example, a source file that contains both object and template literal notation:

```js
const X = () => { return <a style={{ color: "red" }}></a> }
const Y = styled.a`color:red`;
```

See [this demo](https://stylelint.io/demo#N4Igxg9gJgpiBc4IDsDOAXABADUwXkwAoBKfAPk2EwCcZ0BXa5TAHgENMMBPAGxj2BVIPCNXiYAOiFpQpmAL7yyLAPRsK8ickhosATXyd0vGFAB0UAJYA3AAbDR8GbYDcIADQhUXZOjYAPBHBUVABaS2RQgCtUDyRkADNLAHMg4C1MSWl6PlQpcXTmTKzYMB42ajZ0SxRQgCMRMABrUPRKyx4I5NDUGABbSwdkfKy2HgB3Ni48kAyFLXkQeSA).

It's not currently possible to turn on the `declaration-block-trailing-semicolon` rule, and have it ignore the object construct and enforce (a disallow) a semicolon in template literal one.

> What actually happened (e.g. what warnings or errors did you get)?

False positives.

**Comment by jeddy3** [comment]
No longer relevant as [we've frozen the stylistic rules](https://github.com/stylelint/stylelint/issues/5674#issuecomment-1010201282).
